File: dotnet/src/SimpleOrm/ConventionMapLoader.cs
namespace SimpleOrm;

/// <summary>
/// The convention loader (§7.6): for a type with no mapping attributes at all, every
/// public settable property maps by convention; a property named <c>Id</c> is the
/// key (database-generated for int/long, client GUID for Guid). A conventional
/// entity is always a table, named by the convention.
/// </summary>
internal static class ConventionMapLoader
{
    public static EntityMap Load(Type entityType, INamingConvention convention)
    {
        var errors = new List<MappingError>();
        var specs = new List<MappedPropertySpec>();

        foreach (var property in AttributeMapLoader.PropertiesInDeclarationOrder(entityType))
        {
            if (property.GetMethod is not { IsPublic: true } || property.SetMethod is not { IsPublic: true })
            {
                continue;
            }

            var isId = property.Name == "Id";
            specs.Add(new MappedPropertySpec(property)
            {
                IsKey = isId,
                IsGenerated = isId && (property.PropertyType == typeof(int) || property.PropertyType == typeof(long)),
            });
        }

        var map = MapAssembler.Assemble(
            entityType,
            RelationKind.Table,
            convention.TableName(entityType.Name),
            schema: null,
            statementSql: null,
            statementParameters: [],
            specs,
            indexSpecs: [],
            relationshipSpecs: [],
            convention,
            errors);

        if (map is null)
        {
            throw new MappingException(entityType, errors);
        }

        return map;
    }
}

File: php/src/Metadata/ConventionMapLoader.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Metadata;

use ReflectionClass;
use SimpleOrm\Errors\MappingException;
use SimpleOrm\Naming\NamingConvention;

/**
 * The convention loader (§7.6): for a type with no mapping attributes at all,
 * every publicly settable property maps by convention; a property named `id`
 * is the key (database-generated for `int`, natural otherwise). A conventional
 * entity is always a table, named by the convention from the class short name.
 */
final class ConventionMapLoader
{
    private function __construct()
    {
    }

    /** @param class-string $entityType */
    public static function load(string $entityType, NamingConvention $convention): EntityMap
    {
        $class = new ReflectionClass($entityType);
        $errors = [];
        $specs = [];

        foreach (PropertyDiscovery::propertiesInDeclarationOrder($class) as $property) {
            if (!PropertyVisibility::isPubliclySettable($property)) {
                continue;
            }

            $isId = $property->getName() === 'id';
            $resolved = PropertyTypes::resolve($property, null, false);
            $specs[] = new MappedPropertySpec(
                $property,
                $resolved->type,
                $resolved->phpType,
                $resolved->nullable,
                isKey: $isId,
                isGenerated: $isId && $resolved->phpType === 'int',
            );
        }

        $map = MapAssembler::assemble(
            $entityType,
            RelationKind::Table,
            $convention->toDatabase($class->getShortName()),
            schema: null,
            statementSql: null,
            statementParameters: [],
            specs: $specs,
            indexSpecs: [],
            relationshipSpecs: [],
            convention: $convention,
            errors: $errors,
        );

        if ($map === null) {
            throw new MappingException($entityType, $errors);
        }

        return $map;
    }
}
